site/content/blog/2020-05-20-hands-on-iouring-go.md

```diff
@@ -0,0 +1,327 @@
+---
+title: Getting Hands-on with io_uring and Go
+slug: hands-on-iouring-go
+date: 2020-05-20
+categories:
+    - "go"
+author: Agniva De Sarker
+github: agnivade
+community: agnivade
+---
+
+In Linux, syscalls are at the heart of everything. They are the primary interface through which an application interacts with the kernel. Therefore, it is vital that they are fast. And especially in a post-spectre/meltdown world, this is all the more important.
+
+A major chunk of the syscalls deal with I/O, because that's what most applications do. For network I/O, we've had the `epoll` family of syscalls which have provided us with reasonably fast performance. But in the filesystem I/O department, things were a bit lacking. We've had `async_io` for a while now, but apart from a small niche set of applications, it isn't very beneficial. The major reason being that it only works if the file is opened with the `O_DIRECT` flag. This will make the kernel bypass any OS caches and try to read/write from/to the device directly. Not a very good way to do I/O when we are trying to make things go fast. And in buffered mode, it would behave synchronously.
+
+All that is changing slowly, because now we have a brand new interface to perform I/O with the kernel - io_uring. There's a lot of buzz happening around it. And rightly so, because it gives us an entirely new model to interact with the kernel. Let's dive into it and try to understand what it is and how does it solve the problem. And then we will construct a small demo application with Go to play with it.
```

Comment:
```suggestion
All that is changing slowly, because now we have a brand new interface to perform I/O with the kernel - `io_uring`. There's a lot of buzz happening around it. And rightly so, because it gives us an entirely new model to interact with the kernel. Let's dive into it and try to understand what it is and how does it solve the problem. And then we will construct a small demo application with Go to play with it.
```